reuse the digest auth state to avoid unnecessary requests

rettier opened the pull request:
this implements digest authentication state reuse as discussed in #1467.

Contrary to the [comment](https://github.com/encode/httpx/issues/1467#issuecomment-1280656164) i am not skipping most of the authentication flow if a previous challenge exists. Otherwise a failed authentication (401) would not result in a retry with the new server challenge (`test_digest_auth_resets_nonce_count_after_401` would fail).

tomchristie commented on tests/client/test_auth.py:
```diff
@@ -527,6 +527,47 @@ async def test_digest_auth_incorrect_credentials() -> None:
+    auth = DigestAuth(username="tomchristie", password="password123")
+    app = DigestApp()
+
+    async with httpx.AsyncClient(transport=httpx.MockTransport(app)) as client:
+        response_1 = await client.get(url, auth=auth)
+        response_2 = await client.get(url, auth=auth)
+
+        assert response_1.status_code == 200
+        assert response_2.status_code == 200
+
+        assert len(response_1.history) == 1
+        assert len(response_2.history) == 0
```
Like this test, nice and clear. 👍

tomchristie commented on tests/client/test_auth.py:
```diff
@@ -527,6 +527,47 @@ async def test_digest_auth_incorrect_credentials() -> None:
+@pytest.mark.asyncio
+async def test_digest_auth_reuses_challenge() -> None:
+    url = "https://example.org/"
+    auth = DigestAuth(username="tomchristie", password="password123")
```
Eeek... you've included my password in clear text. 😬
I'd suggest anonymising it like this instead...

```suggestion
    auth = DigestAuth(username="user", password="password123")
```

tomchristie commented on tests/client/test_auth.py:
```diff
@@ -527,6 +527,47 @@ async def test_digest_auth_incorrect_credentials() -> None:
+@pytest.mark.asyncio
+async def test_digest_auth_resets_nonce_count_after_401() -> None:
+    url = "https://example.org/"
+    auth = DigestAuth(username="tomchristie", password="password123")
```
```suggestion
    auth = DigestAuth(username="user", password="password123")
```

tomchristie commented on tests/client/test_auth.py:
```diff
@@ -527,6 +527,47 @@ async def test_digest_auth_incorrect_credentials() -> None:
+        first_nonce = response_1.json()["auth"].split("nonce=")[1].split(",")[0]
+        first_nc = response_1.json()["auth"].split("nc=")[1].split(",")[0]
+
+        app.send_response_after_attempt = 2
```
This is a bit obscure to me, I think this line deserves a comment.

rettier replied:
added some explanation for the assertions / send_response_after

tomchristie commented on tests/test_auth.py:
```diff
@@ -61,3 +61,39 @@ def test_digest_auth_with_401():
         flow.send(response)
+
+
+def test_digest_auth_with_401_nonce_counting():
```
Very nice, yup.

tomchristie reviewed (commented):
Great stuff - thanks!
A few minor suggestions here.

rettier commented:
I have refactored some code in view of your suggestions. 
Since you removed your username from the test-data i have also removed it from the existing basic auth test cases.

Regarding the failing test pipeline i assume this is not an issue with my branch, as other pipelines seem to fail with the same error message?

tomchristie commented:
> Regarding the failing test pipeline i assume this is not an issue with my branch, as other pipelines seem to fail with the same error message?

Yup. See https://github.com/encode/httpx/pull/2465

tomchristie commented:
Good work, thanks!
